File: UI_DRC.java
import java.io.*;
import java.net.*;
import org.xbill.DNS.*;
//import Translator.*;
/*
 * This module/library asks questions of given resolver candidates and 
 * reports to texamines answers from a recursive resolver to 
    determine how compliant it is with DNSSEC by asking it a series of 
    question. . 
       Validator
       Partial Validator
       DNSSEC Aware
       Partially DNSSEC Aware 
       Old 
       Not a Resolver 

    The partial systems will have a descriptors reflecting their 
    deficiencies: 
        DNAME:  DNAME processing is not working 
	Permissive: Returns answers that fail validation 
	Mixed: Answers are inconsistent i.e. not all validatable answers are
	       reported as validated. 
	TCP:  TCP queries are not available 
	SlowBig: Big UDP answers fail but TCP fallback is supported
	NoBig:   Both big UDP answers and TCP fail

 */
public class UI_DRC extends DNSSEC_resolver_check {

    static void do_eval(String resolv) {
	String gr = evaluate_resolver(resolv); 
	//print("testing: " + gr);
	String tr = new Translator().translate(gr);
	print("Eval: " + resolv + " " + gr + " " + tr);
    }

    public static void 
    main(String [] args) throws Exception {
	String usage = 
	    "Usage: java -jar DNSSEC-resolver-check.jar -[ahdrS] <resolvers> \n" 
	    + "     :  -a # Aborts on first error for each resolver\n"  
	    + "     :  -d # prints lots of debug info\n"
	    + "     :  -h # prints help and exits\n"  
	    + "     :  -l # Lists the locally configured resolvers and exits\n" 
	    + "     :  -r # detailed report on screen\n"
	    + "     :  -T # Show compact from of test results\n"
	    + "     :  -m # A string that gets added to the report an identifier " 
	    + "or something like that\n\tExample: starbucks\n" 
	    + "     :  -S # DO not SUBMIT results\n" 
	    + "     :    # No resolvers listed, use configured resolvers\n" 
	    + "   resolvers can be addresses or names";
	
	int num_resolvers = 0; 
	boolean resolver_evaluated = false;
	String [] list = ResolverConfig.getCurrentConfig().servers(); 
	int abort = 999999999; 
	
	set_abort(false);
	set_submit_report(true);
	for (num_resolvers = 0; args.length > num_resolvers; num_resolvers++) {
	    if (args[num_resolvers].equals("-a") )
		set_abort(true);
	    else if (args[num_resolvers].equals("-d"))
		set_debug(true);
 	else if (args[num_resolvers].equals("-r"))
	    set_verbose_report(true);
	else if (args[num_resolvers].equals("-S"))
	    set_submit_report(false);
	else if (args[num_resolvers].equals("-T"))
	    set_show_test_results(true); 
	else if (args[num_resolvers].equals("-m")){ 
	    if( num_resolvers + 1 < args.length ) 
		set_message( args[++num_resolvers]);
	    else 
		print( "-m must be followed by a message");
	} else if (args[num_resolvers].equals("-l")) {
	    String msg = "Configured resolvers ";
	    for (int i =0 ; i < list.length; i++)
		msg = msg + " " + list[i]; 
	    print (msg);
	    num_resolvers = abort;
 	} else if (args[num_resolvers].equals("-h")){
	    print(usage);
	    num_resolvers = abort; // abort 
	} else { 
	    resolver_evaluated = true;
	    do_eval(args[num_resolvers]);
	}
	}
	if (resolver_evaluated == false && num_resolvers < abort) 
	    for (int cnt  = 0; cnt < list.length; cnt++) {
		do_eval(list[cnt]);
	    }
	
	System.exit(reports_failed);
    }
}

File: python/UI_DRC.py
# 
# Copyright (c) 2012-2013, Shinkuro. Inc
#
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without 
# modification, are permitted provided that the following conditions are
# met: 
#
#    Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.  
#
#     Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in
#     the documentation and/or other materials provided with the
#     distribution. 
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# AS IS AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 

"""
This is a port of UI_DRC.java (ogud@shinkuro.com) to Python

$Id: UI_DRC.py 446 2013-08-16 20:36:10Z bob.novas $
"""

import dns
import sys
from DNSSEC_resolver_check import DNSSEC_resolver_check 
from Translator import Translator

class UI_DRC: 
   def do_eval(resolv, msg, long_report):
      print resolv, msg
      gr = DNSSEC_resolver_check().evaluate_resolver(resolv, msg)
      tr = Translator().translate(gr)
      if (long_report == True): 
         print "Eval: ", resolv, " Tests=", gr, "Result=", tr
      else:
         print "Result: ", resolv, " ", tr
   def usage():
      use_msg = [ "python UI_DRC.py -[adhlrST] [-m <msg>] [<resolvers>]", 
                  "     :  -a # Aborts on first error for each resolver",
                  "     :  -d # prints lots of debug info",
                  "     :  -h # prints help and exits",
                  "     :  -l # Lists the locally configured resolvers and exits",
                  "     :  -r # detailed report on screen", 
                  "     :  -S # DO not SUBMIT results" ,
                  "     :  -T # Show compact from of test results" ,
                  "     :  -m # A string that gets added to the report an identifier",
                  "\tExample: starbucks",
                  "     :    # No resolvers listed, use the configured resolvers",
                  "\tresolvers can be addresses or names" ]

      for i in use_msg:
         print i

   if __name__ == "__main__": 
      num_resolvers = 0
      resolver_evaluated = False
      abort = False
      long_report = False
      msg = False
      message = "Test-x"  #fix later 
      DRC = DNSSEC_resolver_check();
      list = DRC.get_local_resolvers()
      #       print list
      DRC.set_abort(False);
      DRC.set_submit_report(True)
      for i in sys.argv[1:] :
         print i
         if (i == "-a"):
            DRC.set_abort(True)
         elif (i == "-d"):
            DRC.set_debug(True)
         elif (i == "-r"):
            long_report = True;
            DRC.set_verbose_report(long_report)
         elif (i == "-S"):
            DRC.set_submit_report(False)
         elif (i == "-T"):
            DRC.set_show_test_results(True)
         elif (i == "-m"):
            msg = True
         elif (i== "-l"):
            print "Configured resolvers", list
            abort = True
         elif (i == "-h"):
            usage()
            abort = True
         elif( msg == True):
            message = i = message;
            msg = False
         else:
            resolver_evaluated = True
            print "testing", i
            do_eval( i, message, long_report)

         if(abort == True):
            break;

      if( (resolver_evaluated == False) and (abort == False)):
         for j in list:
            do_eval( j, message, long_report)

